src/provider.rs

```diff
@@ -60,6 +60,33 @@ pub(crate) fn lookup_validator(provider: &str) -> Option<Validator> {
     }
 }
 
+/// The outcome of resolving a provider and making the remote call
+pub enum ProviderOutcome {
+    /// The provider was a [`Resolver`] and returned [`RecordData`].
+    Data(RecordData),
+    /// The provider was a [`Referrer`] and returned a new [`RemoteId`].
+    Reference(RemoteId),
+    /// The provider returned `None`.
+    Null,
+}
+
+/// Obtain the [`ProviderOutcome`] by looking up the [`RemoteId`] using the provided `client`.
+pub fn get_provider_outcome(
+    client: &HttpClient,
+    remote_id: &RemoteId,
+) -> Result<ProviderOutcome, ProviderError> {
+    match lookup_provider(remote_id.provider()) {
+        Either::Left(resolver) => match resolver(remote_id.sub_id(), client)? {
+            Some(data) => Ok(ProviderOutcome::Data(data)),
+            None => Ok(ProviderOutcome::Null),
+        },
+        Either::Right(referrer) => match referrer(remote_id.sub_id(), client)? {
+            Some(new_remote_id) => Ok(ProviderOutcome::Reference(new_remote_id)),
+            None => Ok(ProviderOutcome::Null),
+        },
+    }
+}
+
```

Comment:
Should we actually just expose an entirely different API from the `provider` module? I.e., also do not expose the `lookup_validator` function, and just return another function like `validate_remote_id(&RemoteId) -> bool`. Then the `lookup_validator` and `lookup_provider` are just internal lookup tables for implementation convenience.

Reply:
Sounds reasonable. I don't have any objection.

Reply:
Something I just realized is that the `validate_remote_id` cannot have a call signature which accepts a `&RemoteId`, since by definition once it is a `RemoteId` it is already valid!

But I guess we could just as easily write `validate_remote_id(provider: &str, sub_id: &str) -> bool`. I will implement this.